src/Init/Data/Json/Basic.lean

```diff
@@ -0,0 +1,141 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Gabriel Ebner, Marc Huisinga
+-/
+prelude
+import Init.System.IO
+import Init.Data.String 
+import Init.Data.Int
+
+-- mantissa * 10^-exponent
+structure JsonNumber :=
+(mantissa : Int) (exponent : Nat)
+
+protected def JsonNumber.decEq : ∀ (a b : JsonNumber), Decidable (a = b)
+| ⟨m1, e1⟩, ⟨m2, e2⟩ =>
+  match decEq m1 m2 with
+  | isTrue hm =>
+    match decEq e1 e2 with
+    | isTrue he  => isTrue (hm ▸ he ▸ rfl)
+    | isFalse he => isFalse (fun h => JsonNumber.noConfusion h (fun hm he2 => he he2))
+  | isFalse hm => isFalse (fun h => JsonNumber.noConfusion h (fun hm2 he => hm hm2))
+
+instance JsonNumber.DecidableEq : DecidableEq JsonNumber := 
+JsonNumber.decEq
+
+namespace JsonNumber
+
+protected def fromNat (n : Nat) : JsonNumber := ⟨n, 0⟩
+protected def fromInt (n : Int) : JsonNumber := ⟨n, 0⟩
+
+instance natToJsonNumber : HasCoe Nat JsonNumber := ⟨JsonNumber.fromNat⟩
+instance intToJsonNumber : HasCoe Int JsonNumber := ⟨JsonNumber.fromInt⟩
+
+private def countDigitsAux : Nat → Nat → Nat → Nat
+| 0  , _, _        => 0
+| i+1, n, digits => 
+  if n ≤ 9 then
+    digits
+  else
+    countDigitsAux i (n/10) (digits+1)
+
+private def countDigits (n : Nat) : Nat :=
+countDigitsAux n n 1
+
+protected def toString : JsonNumber → String
+| ⟨m, 0⟩ => m.repr
+| ⟨m, e⟩ =>
+  let s : Bool := m ≥ 0;
+  let m := m.natAbs;
+  -- if there are too many zeroes after the decimal, we
+  -- use exponents to compress the representation.
+  -- this is mostly done for memory usage reasons:
+  -- the size of the representation would otherwise
+  -- grow exponentially in the value of exponent.
+  let exp := 9 + (countDigits m : Int) - (e : Int);
+  let exp := if exp < 0 then exp else 0;
+  let f := 10 ^ (e - exp.natAbs);
+  let left := m / f;
+  let right := (f : Int) + m % f;
+  let rightUntrimmed := right.repr.mkIterator.next.remainingToString;
+  (if s then "" else "-") ++
+  left.repr ++ "." ++
+  (rightUntrimmed.toSubstring.dropRightWhile (fun c => c = '0')).toString ++
+  (if exp = 0 then "" else "e" ++ exp.repr) 
+
+-- shift a JsonNumber by a specified amount of places to the left
+protected def shiftl : JsonNumber → Nat → JsonNumber
+-- if s ≤ e, then 10 ^ (s - e) = 1, and hence the mantissa remains unchanged.
+-- otherwise, the expression pads the mantissa with zeroes
+-- to accomodate for the remaining places to shift. 
+| ⟨m, e⟩, s => ⟨m * (10 ^ (s - e) : Nat), e - s⟩
+
+-- shift a JsonNumber by a specified amount of places to the right
+protected def shiftr : JsonNumber → Nat → JsonNumber
+| ⟨m, e⟩, s => ⟨m, e + s⟩
+
+instance jsonNumberToString : HasToString JsonNumber :=
+⟨JsonNumber.toString⟩
+
+instance jsonNumberHasRepr : HasRepr JsonNumber :=
+⟨fun ⟨m, e⟩ => "⟨" ++ m.repr ++ "," ++ e.repr ++ "⟩"⟩
+
+end JsonNumber
+
+inductive Json
+| null
+| bool (b : Bool)
+| num (n : JsonNumber)
+| str (s : String)
+| arr (elems : List Json)
```

Comment:
@leodemoura Marc realised that the `Array` functions are not universe-polymorphic enough to implement this change, since in `Array.hasFromJson` we need `Array.mapM _ : Array Json -> Option (Array a)` for `Json : Type` and any `a : Type u`, i.e. with different input/output universes. I look into this and told @mhuisi to restrict `Type u` to `Type` for now, but then it turned out to be easier to fix than I assumed: https://github.com/Kha/lean4/commit/4e5fcb6fd774dbc337832d43a17e31235004fb5c
Do you want this fix? Is it correct?

Reply:
I will take a look after the Dev meeting